runtime/src/chpl-qsbr.c

```diff
@@ -0,0 +1,515 @@
+static inline void push_recycle_list(struct tls_node *node, struct defer_node *dnode);
+static inline void push_recycle_list(struct tls_node *node, struct defer_node *dnode) {
+  dnode->next = node->recycleList;
+  node->recycleList = dnode;
+}
+
+// Deletes data held by 'dnode', does not free 'dnode'
+static inline void delete_data(struct defer_node *dnode);
+static inline void delete_data(struct defer_node *dnode) {
+  if (dnode->numData) {
+      void **arr = dnode->data;
+      for (int i = 0; i < dnode->numData; i++) {
+          chpl_mem_free(arr[i], 0, 0);
+      }
+  }
+  chpl_mem_free(dnode->data, 0, 0);
+}
+
+// Initializes TLS; should only need to be called once.
+void chpl_qsbr_init_tls(void);
+void chpl_qsbr_init_tls(void) {
+  struct tls_node *node = chpl_mem_calloc(1, sizeof(struct tls_node), CHPL_RT_MD_QSBR, 0, 0);
+  node->epoch = get_global_epoch();
+
+  // Append to head of list
+  struct tls_node *old_head;
+  do {
+    old_head = chpl_qsbr_tls_list;
+    node->next = old_head;
+  } while (!atomic_compare_exchange_weak_uintptr_t(
+      (uintptr_t *) &chpl_qsbr_tls_list, (uintptr_t) old_head, (uintptr_t) node)
+    );
+
+  CHPL_TLS_SET(chpl_qsbr_tls, node);
+}
+
+// Obtains the minimum epoch of all threads and of the disabled safest epoch.
+static uint64_t safe_epoch(void);
+static uint64_t safe_epoch(void) {
+  uint64_t min = get_disabled_epoch();
+  // Check all remaining threads for whether they have updated to a more recent epoch.
+  for (struct tls_node *node = get_tls_list(); node != NULL; node = node->next) {
+    if (!is_parked(node)) {
+      uint64_t epoch = get_epoch(node);
+      min = (min < epoch) ? min : epoch;       
+    }
+  } 
+  
+  return min;
+}
+
+// Requires spinlock.
+static inline void handle_deferred_data(struct tls_node *node);
+static inline void handle_deferred_data(struct tls_node *node) {
+  // Acquire all data that can be deleted.
+  uint64_t epoch = safe_epoch();
+  struct defer_node *list = pop_bulk_defer_list(node, epoch);
+
+  // Handle deletion of said data...
+  // WARNING: Due to how `--memTrack` works, it is possible for
+  // it to trigger a task-switch when it acquires the memTrack_sync
+  // auxiliary sync var so we must handle this outside of the spinlock.
+  release_spinlock(node);
+  for (struct defer_node *dnode = list; dnode != NULL;) {
+    struct defer_node *tmp = dnode;
+    dnode = dnode->next;
+    delete_data(tmp);
+    push_recycle_list(node, tmp);
+  }
+  acquire_spinlock(node, (uintptr_t) node);
+}
+
+// Requires spinlock for both tls_nodes. Transfers deferred deletion from giver to receiver
+static inline void give_deferred(struct tls_node *giver, struct tls_node *receiver);
+static inline void give_deferred(struct tls_node *giver, struct tls_node *receiver) {
+  struct defer_node *list = pop_all_defer_list(giver);
+  
+  if (receiver->deferList == NULL) {
+    receiver->deferList = list;
+  } else {
+    // Sort list
```

Comment:
Maybe update comment to say "Merge into list sorted in order of decreasing targetEpoch" ?